### PR TITLE
Extract Rebuild library

### PR DESCRIPTION
**Pull request by RikkiGibson**

### CODE REVIEW

**Comment by RikkiGibson** [comment]
I added the new project to the solution via `dotnet sln add` and BuildBoss gives me the following failure:

```
Processing Solution D:\workspace\_work\1\s\Compilers.sln ... FAILED
Processed 71 projects
Project Rebuild.csproj should have GUID 9a19103f-16f7-4668-be54-9a1e7a4f7556 but has fae04ec0-301f-11d3-bf4b-00c04f79efbc

Processing Solution D:\workspace\_work\1\s\Roslyn.sln ... FAILED
Processed 197 projects
Project Rebuild.csproj should have GUID 9a19103f-16f7-4668-be54-9a1e7a4f7556 but has fae04ec0-301f-11d3-bf4b-00c04f79efbc
```

I don't know what this means, but I guess I might need to try again to add the project from Visual Studio. @jaredpar

**Comment by jaredpar** [comment]
> I don't know what this means, but I guess I might need to try again to add the project from Visual Studio.

It just means that the GUID used in the SLN file is incorrect. The line which declares the project , and likely the subsetquent consupmions, use the wrong GUID.

File: src/Compilers/Core/Rebuild/Rebuild.csproj [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
+
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
+    <TargetFrameworks>netcoreapp3.1;netstandard2.0</TargetFrameworks>
+    <PlatformTarget>AnyCPU</PlatformTarget>
+    <Nullable>enable</Nullable>
```
Comment (jaredpar) [inline src/Compilers/Core/Rebuild/Rebuild.csproj:9]: Is this needed? Think it should just come from the `Directory.Build.props` file

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/Core/Rebuild/BuildConstructor.cs [hunk 2]
```diff
@@ -24,23 +26,20 @@
 
 namespace BuildValidator
 {
-    /// <summary>
-    /// An abstraction for building from a MetadataReaderProvider
-    /// </summary>
-    internal class BuildConstructor
+    public class BuildConstructor
```
Comment (cston) [inline src/Compilers/Core/Rebuild/BuildConstructor.cs:29]: >public [](start = 4, length = 6)

Should the types in this assembly be kept as `internal` with `InternalsVisibleTo()` access added for `BuildValidator`?

Comment (RikkiGibson) [reply src/Compilers/Core/Rebuild/BuildConstructor.cs:29]: We do intend for external users to be able to consume this library. I want to punt on the decision of which members to reduce to internal accessibility until we get closer to the point where the public API is ready for review.

File: src/Compilers/Core/Rebuild/Rebuild.csproj (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/Core/Rebuild/Rebuild.csproj:9]: BuildValidator also does this. I'll remove both entries in a subsequent PR.

File: src/Compilers/Core/Rebuild/BuildConstructor.cs (hunk 2 above)
Comment (cston) [inline src/Compilers/Core/Rebuild/BuildConstructor.cs:29]: >class [](start = 11, length = 5)

`sealed`

File: src/Compilers/Core/Rebuild/CompilationOptionsReader.cs [hunk 3]
```diff
@@ -18,34 +18,10 @@
...
-            SourceFilePath = sourceFilePath;
-            HashAlgorithm = hashAlgorithm;
-            Hash = hash;
-            EmbeddedText = embeddedText;
-            EmbeddedCompressedHash = embeddedCompressedHash;
-        }
-    }
-
-    internal class CompilationOptionsReader
+    public class CompilationOptionsReader
```
Comment (cston) [inline src/Compilers/Core/Rebuild/CompilationOptionsReader.cs:24]: >class [](start = 11, length = 5)

`sealed`

File: src/Compilers/Core/Rebuild/BuildConstructor.cs [hunk 4]
```diff
@@ -349,6 +270,66 @@ private static VisualBasicCompilationOptions CreateVisualBasicCompilationOptions
...
+                if (optionsReader.GetMainTypeName() is { } mainTypeName &&
+                    optionsReader.GetMainMethodName() is { } mainMethodName)
+                {
+                    var typeSymbol = producedCompilation.GetTypeByMetadataName(mainTypeName);
+                    if (typeSymbol is object)
+                    {
+                        var methodSymbols = typeSymbol
+                            .GetMembers(mainMethodName)
+                            .OfType<IMethodSymbol>();
+                        return methodSymbols.FirstOrDefault();
```
Comment (cston) [inline src/Compilers/Core/Rebuild/BuildConstructor.cs:327]: >FirstOrDefault [](start = 45, length = 14)

Do we need to handle the case where there are multiple "Main" methods in the type? Could we simply use `producedCompilation.GetEntryPoint()` instead (and avoid looking at `optionsReader` as well)?

Comment (RikkiGibson) [reply src/Compilers/Core/Rebuild/BuildConstructor.cs:327]: I'm interested in looking more closely at this, as I believe it is buggy, however in this PR I will just migrate this as-is.

File: src/Compilers/Core/Rebuild/BuildConstructor.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/Core/Rebuild/BuildConstructor.cs:29]: > We do intend for external users to be able to consume this library.

If these types will be public, we should also review the namespace names when the public API is reviewed.

---
In reply to: [589681272](https://github.com/dotnet/roslyn/pull/51709#discussion_r589681272) [](ancestors = 589681272)

File: src/Tools/BuildValidator/Program.cs [hunk 5]
```diff
@@ -226,39 +235,99 @@ private static bool ValidateFiles(IEnumerable<FileInfo> originalBinaries, BuildC
...
+            var sourceLinkUTF8 = compilationOptionsReader.GetSourceLinkUTF8();
+            if (sourceLinkUTF8 is null)
+            {
+                return default;
+            }
+
+            var parseResult = JsonConvert.DeserializeAnonymousType(Encoding.UTF8.GetString(sourceLinkUTF8), new { documents = (Dictionary<string, string>?)null });
+            var sourceLinks = parseResult.documents.Select(makeSourceLink).ToImmutableArray();
+
+            if (sourceLinks.IsDefault)
```
Comment (cston) [inline src/Tools/BuildValidator/Program.cs:309]: >sourceLinks.IsDefault [](start = 16, length = 21)

Does `ToImmutableArray()` ever return `default`?

**Review by cston — APPROVED** [review]

File: src/Tools/BuildValidator/Program.cs (hunk 5 above)
Comment (RikkiGibson) [reply src/Tools/BuildValidator/Program.cs:309]: I do not think so.

File: src/Compilers/Core/Rebuild/BuildConstructor.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/Core/Rebuild/BuildConstructor.cs:29]: Will address in follow-up.

File: src/Compilers/Core/Rebuild/CompilationOptionsReader.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/Core/Rebuild/CompilationOptionsReader.cs:24]: Will address in follow-up.

**Pull request merged**